### DISCUSSION TITLE
How to find the duration of the step that materialized a particular asset

### DISCUSSION THREAD

**Discussion by sryza** [dagster-io/dagster#12665] [Q&A] [answered]
Inspired by this question on Slack:
> Hi, Is there a way to get the duration of a run in a multi_asset_sensor ? I’m triggering a slack notification job from the sensor and I’d like to add how long did the materialization take. I’m getting most of the information I use from the event log entries from `context.latest_materialization_records_by_key()` but couldn’t find the duration.

**Comment by sryza** [comment, marked as answer]
Here's a function that will do it, given a run, an instance (how to get an instance: https://github.com/dagster-io/dagster/discussions/7427), and an asset key:

```python
from dagster import DagsterInstance, asset, materialize_to_memory, AssetKey


def get_asset_step_duration_seconds(
    run_id: str, asset_key: AssetKey, instance: DagsterInstance
) -> float:
    run_step_stats = instance.get_run_step_stats(run_id)
    for step_stats in run_step_stats:
        if asset_key in [
            event.asset_materialization.asset_key for event in step_stats.materialization_events
        ]:
            return step_stats.end_time - step_stats.start_time

    raise ValueError(f"None of the steps in run {run_id} materialized asset {asset_key}")
```

If you're in a sensor, you can get the instance from `context.instance`.  If you have an `EventLogEntry`, you can get the `run_id` from it.

For a graph-backed asset, this will only return the duration of the step that contains the leaf op that returns the asset, not other ops in the graph that are upstream of it.

**Comment by mkleinbort-ic** [comment]
Hi, i just stumbled onto this. 

Is there a way to add this as part of the asset metadata? Say, as part of the IO manager?
